drivers/gpio/gpio_rzt2m.c

```diff
@@ -211,22 +251,206 @@ static int rzt2m_gpio_configure(const struct device *dev, gpio_pin_t pin, gpio_f
+	if (irq > 13) {
+		return -ENOSYS;
+	}
+
+	if (mode == GPIO_INT_MODE_DISABLED) {
+		rzt2m_gpio_unlock();
+		WRITE_BIT(*pmc_reg, pin, 0);
+
+		/* check if selected pin is using irq line to avoid unregistering other pin irq
+		 * handler
+		 */
+		if (!irq_used_by_other) {
+			rzt2m_gpio_irq_registered_ports[irq].dev = NULL;
+		}
+		rzt2m_gpio_lock();
+		return 0;
+	}
+
+	/* the irq line is used by another pin */
+	if (irq_used_by_other) {
+		return -EBUSY;
+	}
+
+	uint8_t md_mode = 0x0;
+
+	switch (trig) {
+	case GPIO_INT_TRIG_LOW:
+		md_mode = INT_FALLING_EDGE;
+		break;
+	case GPIO_INT_TRIG_HIGH:
+		md_mode = INT_RISING_EDGE;
+		break;
+	case GPIO_INT_TRIG_BOTH:
+		md_mode = INT_BOTH_EDGE;
+	}
+
+	rzt2m_gpio_unlock();
+
+	uint32_t mdx_mask =
+		~((uint32_t)0b11 << irq); /* description of interrupt type has length of 2 bits */
+	ns_portnf_md_val = (ns_portnf_md_val & mdx_mask) | (md_mode << irq);
+	syscon_write_reg(ns_portnf_md_dev, 0, ns_portnf_md_val); /* set interrupt type */
+
+	WRITE_BIT(*pmc_reg, pin, 1); /* enable special function on selected pin */
+
+	/* in case of every pin on every port irq function number is 0 */
+	*pfc_reg &= ~((uint32_t)0b1111 << pin * 4);
+
+	/* register handling interrupt in isr for selected port and pin */
+	rzt2m_gpio_irq_registered_ports[irq].dev = dev;
+	rzt2m_gpio_irq_registered_ports[irq].pin = pin;
+
+	rzt2m_gpio_lock();
+
+	return 0;
+}
+
+static int rzt2m_gpio_manage_callback(const struct device *dev, struct gpio_callback *cb, bool set)
+{
+	struct rzt2m_gpio_data *data = dev->data;
+
+	return gpio_manage_callback(&data->cb, cb, set);
+}
+
 static const struct gpio_driver_api rzt2m_gpio_driver_api = {
 	.pin_configure = rzt2m_gpio_configure,
 	.port_get_raw = rzt2m_gpio_get_raw,
 	.port_set_masked_raw = rzt2m_port_set_masked_raw,
 	.port_set_bits_raw = rzt2m_port_set_bits_raw,
 	.port_clear_bits_raw = rzt2m_port_clear_bits_raw,
-	.port_toggle_bits = rzt2m_gpio_toggle};
+	.port_toggle_bits = rzt2m_gpio_toggle,
+	.pin_interrupt_configure = rzt2m_gpio_pin_interrupt_configure,
+	.manage_callback = rzt2m_gpio_manage_callback};
+
+#define VALUE_2X(i, _) UTIL_X2(i)
+
+#define PIN_IRQ_INITIALIZER(idx, inst)                                                             \
+	COND_CODE_1(DT_INST_PROP_HAS_IDX(inst, irqs, idx),                                         \
+		    ([DT_INST_PROP_BY_IDX(inst, irqs, idx)] =                                      \
+			     DT_INST_PROP_BY_IDX(inst, irqs, UTIL_INC(idx)) + 1,),                \
+		    ())
+
+#define PORT_IRQS_INITIALIZER(inst)                                                                \
+	FOR_EACH_FIXED_ARG(PIN_IRQ_INITIALIZER, (), inst,                                          \
+			   LISTIFY(DT_INST_PROP_LEN_OR(inst, irqs, 0), VALUE_2X, (,)))
+
+#define RZT2M_DECLARE_INIT(inst)                                                                   \
+	COND_CODE_0(inst, (static int rzt2m_gpio_init_##inst(const struct device *dev) \
```

Comment:
The code related to the initialization of common parts of gpio ports (registration of interrupt handlers) that was previously invoked here has been rewritten since your comment. It was moved to the initialization of the one common instance for all ports